tanner/emulators/twig_template_injection.py

```diff
@@ -0,0 +1,77 @@
+import asyncio
+import logging
+
+from tanner.config import TannerConfig
+from tanner.utils.php_sandbox_helper import PHPSandboxHelper
+from tanner.utils import patterns
+
+
+class TwigTemplateInjection:
+    def __init__(self, loop=None):
+        self._loop = loop if loop is not None else asyncio.get_event_loop()
+        self.logger = logging.getLogger("tanner.twig_template_injection")
+        self.helper = PHPSandboxHelper(self._loop)
+        self.autoloader = TannerConfig.get("TWIG_PATH", "autoloader")
+        self.stringloader = TannerConfig.get("TWIG_PATH", "stringloader")
+
+    async def get_injection_result(self, code):
+        """
+        Injects the code from attacker to vulnerable code and get emulation results from php sandbox.
+        :param code (str): Input payload from attacker
+        :return: twig_injection_result (dict): file_md5 (md5 hash), stdout (injection result) as keys.
+        """
+
+        vul_code = """
+        <?php
+
+            require '%s';
+            require '%s';
+
+            Twig_Autoloader::register();
+            $loader = new Twig_Loader_String();
+            $twig = new Twig_Environment($loader);
+            $twig->addExtension(new \\Twig\\Extension\\StringLoaderExtension());
+            $payload = "%s";
+            $result = $twig->render($payload);
+            echo $result;
+        ?>
+        """ % (
+            self.autoloader,
+            self.stringloader,
+            code,
+        )
+
+        self.logger.debug(
+            "Getting the twig injection results of %s from php sandbox", code
+        )
+        twig_injection_result = await self.helper.get_result(vul_code)
+
+        return twig_injection_result
+
+    def scan(self, value):
+        """
+        Scans the input payload to detect attack using regex
+        :param value (str): code from attacker
+        :return: detection (dict): name (attack name), order (attack order) as keys
+        """
+
+        detection = None
+        if patterns.TEMPLATE_INJECTION_TORNADO.match(value):
+            detection = dict(name="twig_template_injection", order=3)
```

Comment:
I tested with super simple string /foobar?b={{3*%272%27}} There are 2 possible detections:
```
{'name': 'template_injection', 'order': 4}
{'name': 'twig_template_injection', 'order': 3}
```
Since `template_injection` is order 4, tanner chooses it and execute. 

Any ideas how to distinguish? @mzfr @rjt-gupta

Reply:
For twig this case should give `6` and for tornado `222`.

Reply:
we need something that distinguishes these two..

Reply:
yeah the regex is almost similar so it' hard to distinguish.

Also afeena can you please tell me how did you tested it? I mean the setup I want to know what I am doing wrong.

Reply:
@mzfr scan works perfectly fine, if you print possible detection you can see it. the only problem in the signature of the `handle` function.